scipy/_lib/_array_api.py

```diff
@@ -561,3 +570,41 @@ def xp_take_along_axis(arr: Array,
         raise NotImplementedError("Array API standard does not define take_along_axis")
     else:
         return xp.take_along_axis(arr, indices, axis)
+
+
+# utility to broadcast arrays and promote to common dtype
+def xp_broadcast_promote(*args, ensure_writeable=False, force_floating=False, xp=None):
+    xp = array_namespace(*args) if xp is None else xp
+
+    args = [(xp.asarray(arg) if arg is not None else arg) for arg in args]
+    args_not_none = [arg for arg in args if arg is not None]
+
+    # determine minimum dtype
+    dtypes = [arg.dtype for arg in args_not_none
+              if not xp.isdtype(arg.dtype, 'integral')]
+    if force_floating:
+        dtypes.append(xp.asarray(1.).dtype)
+    dtype = xp.result_type(*dtypes)
```

Comment:
This is typically closer to what we do. Not sure why I did something different here.
```suggestion
    dtypes = [arg.dtype for arg in args_not_none]
    dtype = xp.result_type(*dtypes)
    if force_floating and xp.isdtype(dtype, 'integral'):
        dtype = xp.asarray(1.).dtype
```